packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx

```diff
@@ -4,171 +4,273 @@
-          isPlanProjection ? acc.models2d.push(id) : acc.models3d.push(id);
-          return acc;
-        }, { models2d: [] as string[], models3d: [] as string[] });
-
-        setAvailable2dModels(models2d);
-        setAvailable3dModels(models3d);
+      .then((modelInfos: ModelInfo[]) => {
+        setAvailableModels(modelInfos);
       })
       .catch((_e) => {
         setAvailableModels([]);
       });
   }, [queryModels, viewport]);
 
-  const invert = useCallback(async () => {
-    if (availableModels.length === 0) return;
-    const notViewedModels: string[] = [];
-    const models: string[] = [];
-    availableModels.forEach((id: string) => {
-      if (viewport?.viewsModel(id)) models.push(id);
-      else notViewedModels.push(id);
-    });
-    await viewport?.addViewedModels(notViewedModels);
-    viewport?.changeModelDisplay(models, false);
-    viewport?.invalidateScene();
-  }, [viewport, availableModels]);
-
-  const hideAll = useCallback(async () => {
-    viewport?.changeModelDisplay(availableModels, false);
-    viewport?.clearAlwaysDrawn();
-    if (iModel) {
+  return (
+    <>
+      <SearchBar
+        value=""
+        valueChangedDelay={500}
+        placeholder={TreeWidget.translate("search")}
+        title={TreeWidget.translate("searchForSomething")}
+        filteringInProgress={searchOptions.isFiltering}
+        onFilterCancel={searchOptions.onFilterCancel}
+        onFilterClear={searchOptions.onFilterCancel}
+        onFilterStart={searchOptions.onFilterStart}
+        onSelectedChanged={searchOptions.onResultSelectedChanged}
+        resultCount={searchOptions.matchedResultCount ?? 0}
+      >
+        {props.headerButtons
+          ? props.headerButtons.map((btn, index) =>
+            <React.Fragment key={index}>
+              {btn({ viewport, models: availableModels })}
+            </React.Fragment>)
+          : [
+            <ShowAllButton viewport={viewport} models={availableModels} key="show-all-btn" />,
+            <HideAllButton viewport={viewport} models={availableModels} key="hide-all-btn" />,
+            <InvertButton viewport={viewport} models={availableModels} key="invert-all-btn" />,
+            <View2DButton viewport={viewport} models={availableModels} key="view-2d-btn" />,
+            <View3DButton viewport={viewport} models={availableModels} key="view-3d-btn" />,
+          ]
+        }
+      </SearchBar>
+      <AutoSizer>
+        {({ width, height }) => (
+          <ModelsTree
+            {...props}
+            iModel={iModel}
+            activeView={viewport}
+            width={width}
+            height={height}
+            filterInfo={{ filter: filterString, activeMatchIndex }}
+            onFilterApplied={onFilterApplied}
+          />
+        )}
+      </AutoSizer>
+    </>
+  );
+}
+
+function ShowAllButton(props: ModelsTreeHeaderButtonProps) {
+  const showAll = async () => {
+    if (!props.models || !props.viewport)
```

Comment:
Looking at `ModelsTreeComponentImpl` it looks like `models` and `viewport` are always defined.

Changing `ModelsTreeHeaderButtonProps` to have non-optional `models` propety would allow to remove quite some `undefined` checks `?` in button implementations.

Reply:
fixed